### PR TITLE
services.xserver.imwheel: add module

### PR DESCRIPTION
**Pull request by turboMaCk**
Add [IMWheel](https://wiki.archlinux.org/index.php/IMWheel) module.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Comment by turboMaCk** [comment]
rebased. Any idea who to ask for a review @worldofpeace @Infinisil @jhillyerd ?

**Comment by jhillyerd** [comment]
I haven't done services yet with NixOS, so I'm probably not the best to review.  One comment though: could we make `-d` mandatory instead of default, so that imwheel never daemonizes itself, and drop the ExecStop=pkill line?

**Comment by turboMaCk** [comment]
I guess that would be an option though I'm not sure if there is not any valid usecases for overriding that flug though I guess there isn't any.

**Comment by turboMaCk** [comment]
@jhillyerd I think I found a problem with this. If `-d` would be present in options this would not work. You can try to run `imwheel -d -d` which will start imwheel as a deamon. Same way `imwheel -d -d -d` works same way as `-d` so there is probably some negation going on. I personally feel that theit would be better to keep this as is but it's open to discussion.

File: nixos/modules/services/x11/imwheel.nix [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+          '';
+          description = "Configuration";
+        };
+      };
+    };
+
+    config = mkIf cfg.enable {
+      environment.systemPackages = with pkgs; [ imwheel ];
+
+      nixpkgs.overlays = singleton (self: super: {
```
Comment (worldofpeace) [inline nixos/modules/services/x11/imwheel.nix]: Why is this needed? Why does it read the rcfile from within its nix store path and not `/etc/X11/imwheel/imwheelrc`?
I see there's a patch here
```
substituteInPlace Makefile.in --replace "ETCDIR = " "ETCDIR = $out"
substituteInPlace util.c --replace "/etc/X11/imwheel" "$out/etc/X11/imwheel"
```

Which works on `util.c`, we should probably remove this and perhaps use `installFlags` to make it install the file to the nix store.

Comment (turboMaCk) [reply nixos/modules/services/x11/imwheel.nix]: yes that would be an option but it requires change in the definition of the package.

Comment (worldofpeace) [reply nixos/modules/services/x11/imwheel.nix]: Correct, I believe this change is needed for this to function correctly.
All packages that read a config from /etc in nixpkgs are compiled with sysconfdir = /etc and if they have a configuration in the package installed to `$out/etc`. This is because they're configured with a nixos module.

Comment (turboMaCk) [reply nixos/modules/services/x11/imwheel.nix]: sory it took so long @worldofpeace I needed to find time to fiddel with this. Please check if this is what you've ment. No need to hurry though.

File: nixos/modules/services/x11/imwheel.nix [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  cfg = config.services.xserver.imwheel;
+in
+  {
+    options = {
+      services.xserver.imwheel = {
+        enable = mkEnableOption "IMWheel service";
+
+        options = mkOption {
+          type = types.listOf types.str;
+          default = ["-d" "-b 45" "--kill"];
```
Comment (jhillyerd) [inline nixos/modules/services/x11/imwheel.nix]: nit: why mix long and short argument forms?

File: pkgs/tools/X11/imwheel/default.nix [hunk 3]
```diff
@@ -10,10 +10,7 @@ stdenv.mkDerivation rec {
 
   buildInputs = [ libX11 libXext libXi libXmu libXt libXtst ];
 
-  postPatch = ''
-    substituteInPlace Makefile.in --replace "ETCDIR = " "ETCDIR = $out"
-    substituteInPlace util.c --replace "/etc/X11/imwheel" "$out/etc/X11/imwheel"
-  '';
+  installFlags = ["sysconfdir=/etc" "ETCDIR=/etc" ];
```
Comment (jhillyerd) [inline pkgs/tools/X11/imwheel/default.nix]: nit: inconsistent spacing around brackets

**Review by jhillyerd — APPROVED** [review]
approving w/ a couple nit picks.

File: nixos/modules/services/x11/imwheel.nix (hunk 2 above)
Comment (turboMaCk) [reply nixos/modules/services/x11/imwheel.nix]: good poit, just a negligence on my part. I would use long forms as those are more descriptive, agree?

Comment (jhillyerd) [reply nixos/modules/services/x11/imwheel.nix]: I like long form also.

File: pkgs/tools/X11/imwheel/default.nix (hunk 3 above)
Comment (worldofpeace) [reply pkgs/tools/X11/imwheel/default.nix]: I believe you should do this
```
  makeFlags = [
    "sysconfdir=/etc"
    "ETCDIR=/etc"
  ];

  installFlags = [
    "sysconfdir=${placeholder "out"}/etc"
    "ETCDIR=${placeholder "out"}/etc"
  ];
```

So when things are compiled, I hope they get that value because of the `makeFlags`.
But for installation they still end up in `$out`. With this it fails with a permission denied error, which is weird because it really should fail the build entirely (since a step in the phase failed...).

Comment (worldofpeace) [reply pkgs/tools/X11/imwheel/default.nix]: Also, I checked the source for that patched entry and it's
```
	char fname[2][1024]={"","/etc/X11/imwheel/imwheelrc"}, line[1024], *p, *q, winid[1024];
```

So it's already hardcoded, so disregard my point about "compiled" as there's no substitution going on here.
However this is still needed so it can read the config from the dir `/etc/X11/imwheel/imwheelrc`

File: nixos/modules/services/x11/imwheel.nix [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            Additional command-line arguments to pass to
+            <command>imwheel</command>.
+          '';
+        };
+
+        rules = mkOption {
+          type = types.attrsOf types.str;
+          default = {};
+          example = literalExample ''
+            ".*" = \'\'
```
Comment (Infinisil) [inline nixos/modules/services/x11/imwheel.nix]: This evaluates to literally `\'\'`. If you want to escape `''` you need to use `'''` (three single ones)

File: nixos/modules/services/x11/imwheel.nix [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+              None,      Up,   Button4, 8
+              None,      Down, Button5, 8
+              Shift_L,   Up,   Shift_L|Button4, 4
+              Shift_L,   Down, Shift_L|Button5, 4
+              Control_L, Up,   Control_L|Button4
+              Control_L, Down, Control_L|Button5
+            \'\';
+          '';
+          description = "Configuration";
```
Comment (Infinisil) [inline nixos/modules/services/x11/imwheel.nix]: Description should be expanded a bit to form a sentence, and I recommend including a link to upstream docs too.

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/services/x11/imwheel.nix [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+              Shift_L,   Down, Shift_L|Button5, 4
+              Control_L, Up,   Control_L|Button4
+              Control_L, Down, Control_L|Button5
+            ''';
+          '';
+          description = '''
+            Window's translation rules.
+            /etc/X11/imwheelrc is generated based on this config
+            which means this config is global for all users.
+            See http://imwheel.sourceforge.net/imwheel.1.html for more information.
```
Comment (turboMaCk) [inline nixos/modules/services/x11/imwheel.nix]: Honestly the best documentation is probably ArchWiki, anyway it feels more right to point to the official documentation.

File: nixos/modules/services/x11/imwheel.nix [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+              None,      Up,   Button4, 8
+              None,      Down, Button5, 8
+              Shift_L,   Up,   Shift_L|Button4, 4
+              Shift_L,   Down, Shift_L|Button5, 4
+              Control_L, Up,   Control_L|Button4
+              Control_L, Down, Control_L|Button5
+            ''';
+          '';
+          description = '''
+            Window's translation rules.
```
Comment (turboMaCk) [inline nixos/modules/services/x11/imwheel.nix]: I don't honestly even know what is the right terminology to use there. Any suggestion for improvements are welcome.

**Comment by turboMaCk** [comment]
changes based on feedback are implemented but I would still want to try it out on my desktop to be sure it works.

File: nixos/modules/services/x11/imwheel.nix [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            ".*" = '''
+              None,      Up,   Button4, 8
+              None,      Down, Button5, 8
+              Shift_L,   Up,   Shift_L|Button4, 4
+              Shift_L,   Down, Shift_L|Button5, 4
+              Control_L, Up,   Control_L|Button4
+              Control_L, Down, Control_L|Button5
+            ''';
+          '';
+          description = '''
```
Comment (Infinisil) [inline nixos/modules/services/x11/imwheel.nix]: No need for triple quotes here, they're only needed when you want to escape `''` in a `''` string

File: nixos/modules/services/x11/imwheel.nix (hunk 6 above)
Comment (Infinisil) [reply nixos/modules/services/x11/imwheel.nix]: Sounds good, use the url syntax though: `<link xlink:href="https://example.com"/>`

**Comment by turboMaCk** [comment]
Ready for next review. Sorry it took so long. I needed to find some time while I'm on a machine that actually has mouse attached. Example configuration I've tested this with:

```
  services.xserver.imwheel = {
    enable = true;
    rules = {
      "chromium-browser" = ''
        None,      Up,   Button4, 8
        None,      Down, Button5, 8
        Shift_L,   Up,   Shift_L|Button4, 4
        Shift_L,   Down, Shift_L|Button5, 4
        Control_L, Up,   Control_L|Button4
        Control_L, Down, Control_L|Button5
      '';
    };
  };
```

I've also tested 2 months with multiple buttons to make sure default configuration doesn't break things like back and forward buttons.

File: nixos/modules/services/x11/imwheel.nix [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  cfg = config.services.xserver.imwheel;
+in
+  {
+    options = {
+      services.xserver.imwheel = {
+        enable = mkEnableOption "IMWheel service";
+
+        options = mkOption {
+          type = types.listOf types.str;
+          default = ["--detach" "--buttons 45" "--kill"];
```
Comment (Infinisil) [inline nixos/modules/services/x11/imwheel.nix]: I'd do some changes regarding this:
- If users set this to e.g. `["--buttons 45"]` they would override the `--detach` option (which actually should be called `--no-detach` or so anyways), so the service would fork off, which is not as nicely handled anymore then. So I suggest to add `--detach` unconditionally in the `ExecStart` to prevent this. I don't see any reason anybody would want to change that anyways.
- The name `options` is already an overloaded term from the module system, and with above change, `extraOptions` is more fitting, which is also the standard naming for such options.
- Unless I'm missing something, I don't see a reason `--buttons 45` should be passed by default. Same with `--kill` which doesn't make any sense here, or does it? So I'd just make `default = []` but put some examples in `example = [ "--debug" "--buttons 45" ]`

Comment (turboMaCk) [reply nixos/modules/services/x11/imwheel.nix]: These defaults are based mostly on my experimentation because sadly running IMWheel without flags causes some issues

- `detach` is by default **true** so passing `--detach` toggles it so **it doesn't detach**. Which I believe makes more sense when it's ran as a service.
- `kill` meas that any already running IMWheel process will be killed when this one starts. I think this also make sense because otherwise result can be unpredictable
- `buttons 45` is passed because without this IMWheel breaks functionality of additional mouse buttons like back and forward (tested with Logitech and MAD Rat)

Using defaults make it still possible for user to easily overwrite those things. This is why there is a `pkill` https://github.com/NixOS/nixpkgs/pull/71052/files#diff-ea25b843a6681ed1b7711805ec769667R64 so in case `--detach` is not passed (and so process is detached) IMWheel is still killed when service stops. This is also related to this comment https://github.com/NixOS/nixpkgs/pull/71052#issuecomment-544270064 and my reaction https://github.com/NixOS/nixpkgs/pull/71052#issuecomment-544281927.

so TL;DR:

- `--detach` is to prevent detaching which imo makes more sense for a service
- `--buttons 45` is to avoid breaking addition mouse buttons with default settings (just enable)
- `--kill` is to avoid multiple IMWheel processes
- `pkill` is used in cases someone overwrides defaults but doesn't pass `--detach` or `-d`

Which this in mind do you think want to change these defaults @Infinisil ?

Comment (turboMaCk) [reply nixos/modules/services/x11/imwheel.nix]: sorry for ping @Infinisil I don't want to be too pushy about this. I'm happy to implement any changes, just want to make sure you still feel we want to do them considering the comment above.

**Comment by turboMaCk** [comment]
rebased onto master

File: nixos/modules/services/x11/imwheel.nix (hunk 7 above)
Comment (jhillyerd) [reply nixos/modules/services/x11/imwheel.nix]: Maybe `Window class translation rules.` since that's what is being match?

File: nixos/modules/services/x11/imwheel.nix (hunk 9 above)
Comment (jhillyerd) [reply nixos/modules/services/x11/imwheel.nix]: It seems regardless of any changes to the defaults, the rename from `options` to `extraOptions` should happen.

Comment (Infinisil) [reply nixos/modules/services/x11/imwheel.nix]: Hm okay I can see the `buttons` default being reasonable, but `--detach` and maybe `--kill` too can imo be added like
```nix
{
  ExecStart = "${pkgs.imwheel}/bin/imwheel" + escapeShellArgs ([
    "--detach"
    "--kill"
  ] ++ cfg.extraOptions);
}
```

This means even when the user doesn't specify them in `extraOptions`, they still get added, and the expectation is that users won't add `--detach` in `extraOptions`. It should feel like "Extra options in addition to the ones NixOS adds on its own because they always make sense to use"

Comment (turboMaCk) [reply nixos/modules/services/x11/imwheel.nix]: Thanks @Infinisil for the feedback. I've incorporate this change. just few remarks:

- I've add space behind `bin/imwheel` because I don't think escapeShellArgs do that. Let me know if that's not right.
- There is still the pkill on `ExecStop`. With these additional changes it might make sense to incorporate even [change to Simple type](https://github.com/NixOS/nixpkgs/pull/71052#issuecomment-544270064) - depends of if we want to try to protect user from shooting him/herself in foot (multiple --detach arguments case) or we want to make assumption default is good and if someone overrides it by compination of extraOptions then it's his/her thing to deal with.

File: nixos/modules/services/x11/imwheel.nix (hunk 7 above)
Comment (turboMaCk) [reply nixos/modules/services/x11/imwheel.nix]: Thanks, I like it

File: nixos/modules/services/x11/imwheel.nix [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        pkgs.writeText "imwheelrc" (concatStringsSep "\n\n"
+          (mapAttrsToList
+            (rule: conf: "\"${rule}\"\n${conf}") cfg.rules
+          ));
+
+      systemd.user.services.imwheel = {
+        description = "imwheel service";
+        wantedBy = [ "graphical-session.target" ];
+        partOf = [ "graphical-session.target" ];
+        serviceConfig = with lib; {
```
Comment (Infinisil) [inline nixos/modules/services/x11/imwheel.nix]: Just a small thing: You already `with lib;`'d above, so this isn't needed again here

Comment (turboMaCk) [reply nixos/modules/services/x11/imwheel.nix]: good catch! fixed in last commit.

**Review by infinisil — APPROVED** [review]
Looking good!
